core/src/main/java/io/grpc/ExponentialBackoffPolicy.java

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * Exponential backoff policy for reconnects.
+ */
+public final class ExponentialBackoffPolicy implements BackoffPolicy {
+  /**
+   * Provider tuned for connection retries.
+   *
+   * https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md
+   */
+  public static final class Provider implements BackoffPolicy.Provider {
+    @Override
+    public BackoffPolicy get() {
+      return ExponentialBackoffPolicy.builder()
+        .initialBackoffMilis(TimeUnit.SECONDS.toMillis(1))
+        .maxBackoffMillis(TimeUnit.MINUTES.toMillis(2))
+        .multiplier(1.6)
+        .jitter(.2)
+        .build();
+    }
+  }
+
+  private final Random random;
+  private final long initialBackoffMillis;
+  private final long maxBackoffMillis;
+  private final double multiplier;
+  private final double jitter;
+
+  private long nextBackoffMillis;
+
+  private ExponentialBackoffPolicy(Random random, long initialBackoffMilis, long maxBackoffMillis,
+                                   double multiplier, double jitter) {
+    this.random = random;
+    this.initialBackoffMillis = initialBackoffMilis;
+    this.maxBackoffMillis = maxBackoffMillis;
+    this.multiplier = multiplier;
+    this.jitter = jitter;
+
+    this.nextBackoffMillis =  initialBackoffMillis;
+  }
+
+  @Override
+  public long nextBackoffMillis() {
+    long currentBackoffMillis = nextBackoffMillis;
+    nextBackoffMillis = Math.min((long) (currentBackoffMillis * multiplier), maxBackoffMillis);
+    return currentBackoffMillis
+        + uniformRandom(-jitter * currentBackoffMillis, jitter * currentBackoffMillis);
+  }
+
+  private long uniformRandom(double low, double high) {
+    checkArgument(high >= low);
+    double mag = high - low;
+    return (long) (random.nextDouble() * mag + low);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    private Random random;
+    private long initialBackoffMilis;
+    private long maxBackoffMillis;
+    private double multiplier;
+    private double jitter;
+
+    private Builder() {
+    }
+
+    public Builder random(Random random) {
+      checkNotNull(random);
+      return this;
+    }
+
+    public Builder initialBackoffMilis(long initialBackoffMilis) {
+      checkState(initialBackoffMilis > 0, "initialBackoffMilis must be greater than zero");
+      this.initialBackoffMilis = initialBackoffMilis;
+      return this;
+    }
+
+    public Builder maxBackoffMillis(long maxBackoffMillis) {
+      checkState(maxBackoffMillis > 0, "maxBackoffMillis must be greater than zero");
+      this.maxBackoffMillis = maxBackoffMillis;
+      return this;
+    }
+
+    public Builder multiplier(double multiplier) {
+      checkState(multiplier > 0, "multiplier must be greater than zero");
+      this.multiplier = multiplier;
+      return this;
+    }
+
+    public Builder jitter(double jitter) {
+      checkState(jitter >= 0, "jitter must be greater or equal zero");
```

Comment:
checkState is for checking if methods are called out of order, like calling file.read() after file.close()  (The object itself is not in a state that can support the method invocation).

Use checkArgument instead